### PR TITLE
Add OpenSearch mixin

### PR DESCRIPTION
**Pull request by JonathanWamsley**
Adds 2 dashboards and several new alerts for OpenSearch.

Dashboards:
- OpenSearch cluster overview
- OpenSearch node overview
- OpenSearch search and index overview

Alerts:
- OpenSearchYellowCluster
- OpenSearchRedCluster
- OpenSearchUnstableShardReallocation
- OpenSearchUnstableShardUnassigned
- OpenSearchModerateNodeDiskUsage
- OpenSearchHighNodeDiskUsage
- OpenSearchModerateNodeCPUUsage
- OpenSearchHighNodeCPUUsage
- OpenSearchModerateNodeMemoryUsage
- OpenSearchHighNodeMemoryUsage
- OpenSearchModerateRequestLatency
- OpenSearchModerateIndexLatency

## OpenSearch cluster overview

Screenshot 2023-04-24 at 4 43 32 PM

Screenshot 2023-04-24 at 4 43 44 PM

## OpenSearch node overview

Screenshot 2023-04-24 at 4 44 03 PM

Screenshot 2023-04-24 at 4 44 48 PM

## OpenSearch search and index overview

Screenshot 2023-04-24 at 4 45 20 PM

Screenshot 2023-04-24 at 4 45 34 PM

### CODE REVIEW

File: opensearch-mixin/.lint [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  template-datasource-rule:
+    reason: "Based on new convention we are using variable names prometheus_datasource and loki_datasource where as linter expects 'datasource'"
+  template-instance-rule:
+    reason: "Based on new convention we are using variable names prometheus_datasource and loki_datasource where as linter expects 'datasource'"
+  target-instance-rule:
+    reason: "The dashboard monitors on a cluster, node or index level, not an instance level."
+    entries:
+    - dashboard: "OpenSearch cluster overview"
+    - dashboard: "OpenSearch node overview"
+    - dashboard: "OpenSearch search and index overview"
```
Comment (kuiperda) [inline opensearch-mixin/.lint:15]:
```suggestion
    - dashboard: "OpenSearch search and index overview"
  panel-units-rule:
    reason: "Custom units are used for better user experience in these panels"
    entries:
    - panel: (panel name)
```

run `make fmt` and then `make lint` which will give errors for panels that have custom units or no units. 
Should add each panel that this affects under `entries`

**Review by kuiperda — APPROVED** [review]
Looks good!

File: opensearch-mixin/README.md [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+
+```
+{
+  _config+:: {
+    enableLokiLogs: false,
+  },
+}
+```
+
+In order for the selectors to properly work for system logs ingested into your logs datasource, please also include the matching `instance`, `job`, and `cluster` labels onto the [scrape_configs](https://grafana.com/docs/loki/latest/clients/promtail/configuration/#scrape_configs) as to match the labels for ingested metrics.
```
Comment (StefanKurek) [inline opensearch-mixin/README.md:46]:
```suggestion
In order for the selectors to properly work for system logs ingested into your logs datasource, please also include the matching `job` and `opensearch_cluster` labels onto the [scrape_configs](https://grafana.com/docs/loki/latest/clients/promtail/configuration/#scrape_configs) as to match the labels for ingested metrics.
```

File: opensearch-mixin/README.md [hunk 3]
```diff
@@ -0,0 +1,131 @@
...
+
+```yaml
+scrape_configs:
+  - job_name: integrations/opensearch
+    static_configs:
+      - targets: [localhost]
+        labels:
+          job: integrations/opensearch
+          instance: '<your-instance-name>'
```
Comment (StefanKurek) [inline opensearch-mixin/README.md:55]:
```suggestion
```

File: opensearch-mixin/README.md [hunk 4]
```diff
@@ -0,0 +1,131 @@
...
+OpenSearchModerateRequestLatency: The request latency has exceeded the warning threshold.
+OpenSearchModerateIndexLatency: The index latency has exceeded the warning threshold.
+
+Default thresholds can be configured in `config.libsonnet`
+
+```js
+{
+  _config+:: {
+    alertsWarningShardReallocations: 0,
+    alertsWarningShardUnassigned: 1,
```
Comment (StefanKurek) [inline opensearch-mixin/README.md:89]:
```suggestion
    alertsWarningShardUnassigned: 0,
```

File: opensearch-mixin/alerts/alerts.libsonnet [hunk 5]
```diff
@@ -0,0 +1,205 @@
+{
+  prometheusAlerts+:: {
+    groups+: [
+      {
+        name: 'opensearch',
+        rules: [
+          {
+            alert: 'OpenSearchYellowCluster',
+            expr: |||
+              max by (job, cluster) (opensearch_cluster_status) == 1
```
Comment (StefanKurek) [inline opensearch-mixin/alerts/alerts.libsonnet:10]:
```suggestion
              opensearch_cluster_status == 1
```

File: opensearch-mixin/alerts/alerts.libsonnet [hunk 6]
```diff
@@ -0,0 +1,205 @@
...
+              description:
+                (
+                  '{{$labels.opensearch_cluster}} health status is yellow over the last 5 minutes'
+                ) % $._config,
+            },
+          },
+          {
+            alert: 'OpenSearchRedCluster',
+            expr: |||
+              max by (job, cluster) (opensearch_cluster_status) == 2
```
Comment (StefanKurek) [inline opensearch-mixin/alerts/alerts.libsonnet:27]:
```suggestion
              opensearch_cluster_status == 2
```

File: opensearch-mixin/alerts/alerts.libsonnet [hunk 7]
```diff
@@ -0,0 +1,205 @@
...
+              description:
+                (
+                  '{{$labels.opensearch_cluster}} health status is red over the last 5 minutes'
+                ) % $._config,
+            },
+          },
+          {
+            alert: 'OpenSearchUnstableShardReallocation',
+            expr: |||
+              sum by(job, cluster) (opensearch_cluster_shards_number{type="relocating"}) > %(alertsWarningShardReallocations)s
```
Comment (StefanKurek) [inline opensearch-mixin/alerts/alerts.libsonnet:44]:
```suggestion
              sum without(type) (opensearch_cluster_shards_number{type="relocating"}) > %(alertsWarningShardReallocations)s
```

File: opensearch-mixin/config.libsonnet [hunk 8]
```diff
@@ -0,0 +1,22 @@
+{
+  _config+:: {
+    dashboardTags: ['opensearch-mixin'],
+    dashboardPeriod: 'now-1h',
+    dashboardTimezone: 'default',
+    dashboardRefresh: '1m',
+
+    // alerts thresholds
+    alertsWarningShardReallocations: 0,
+    alertsWarningShardUnassigned: 1,
```
Comment (StefanKurek) [inline opensearch-mixin/config.libsonnet:10]:
```suggestion
    alertsWarningShardUnassigned: 0,
```

File: opensearch-mixin/alerts/alerts.libsonnet [hunk 9]
```diff
@@ -0,0 +1,205 @@
...
+              summary: 'A node has gone offline or has been disconnected triggering shard reallocation.',
+              description: |||
+                {{$labels.opensearch_cluster}} has had {{ printf "%%.0f" $value }} shard reallocation over the last 1m which is above the threshold of %(alertsWarningShardReallocations)s.
+              ||| % $._config,
+            },
+          },
+          {
+            alert: 'OpenSearchUnstableShardUnassigned',
+            expr: |||
+              sum by(job, cluster) (opensearch_cluster_shards_number{type="unassigned"}) > %(alertsWarningShardUnassigned)s
```
Comment (StefanKurek) [inline opensearch-mixin/alerts/alerts.libsonnet:60]:
```suggestion
              sum without(type) (opensearch_cluster_shards_number{type="unassigned"}) > %(alertsWarningShardUnassigned)s
```

File: opensearch-mixin/alerts/alerts.libsonnet [hunk 10]
```diff
@@ -0,0 +1,205 @@
...
+              summary: 'There are shards that have been detected as unassigned.',
+              description: |||
+                {{$labels.opensearch_cluster}} has had {{ printf "%%.0f" $value }} shard unassigned over the last 5m which is above the threshold of %(alertsWarningShardUnassigned)s.
+              ||| % $._config,
+            },
+          },
+          {
+            alert: 'OpenSearchModerateNodeDiskUsage',
+            expr: |||
+              100 * sum by(node, job, cluster) ((opensearch_fs_path_total_bytes - opensearch_fs_path_free_bytes) / opensearch_fs_path_total_bytes) > %(alertsWarningDiskUsage)s
```
Comment (StefanKurek) [inline opensearch-mixin/alerts/alerts.libsonnet:76]:
```suggestion
              100 * sum without(nodeid, path, mount, type) ((opensearch_fs_path_total_bytes - opensearch_fs_path_free_bytes) / opensearch_fs_path_total_bytes) > %(alertsWarningDiskUsage)s
```

File: opensearch-mixin/alerts/alerts.libsonnet [hunk 11]
```diff
@@ -0,0 +1,205 @@
...
+              summary: 'The node disk usage has exceeded the warning threshold.',
+              description: |||
+                {{$labels.opensearch_node}} has had {{ printf "%%.0f" $value }} disk usage over the last 5m which is above the threshold of %(alertsWarningDiskUsage)s.
+              ||| % $._config,
+            },
+          },
+          {
+            alert: 'OpenSearchHighNodeDiskUsage',
+            expr: |||
+              100 * sum by(node, job, cluster) ((opensearch_fs_path_total_bytes - opensearch_fs_path_free_bytes) / opensearch_fs_path_total_bytes) > %(alertsCriticalDiskUsage)s
```
Comment (StefanKurek) [inline opensearch-mixin/alerts/alerts.libsonnet:92]:
```suggestion
              100 * sum without(nodeid, path, mount, type) ((opensearch_fs_path_total_bytes - opensearch_fs_path_free_bytes) / opensearch_fs_path_total_bytes) > %(alertsCriticalDiskUsage)s
```

**Review by StefanKurek — CHANGES_REQUESTED** [review]

File: opensearch-mixin/alerts/alerts.libsonnet [hunk 12]
```diff
@@ -0,0 +1,205 @@
...
+              summary: 'The node disk usage has exceeded the critical threshold.',
+              description: |||
+                {{$labels.opensearch_node}} has had {{ printf "%%.0f" $value }}%% disk usage over the last 5m which is above the threshold of %(alertsCriticalDiskUsage)s.
+              ||| % $._config,
+            },
+          },
+          {
+            alert: 'OpenSearchModerateNodeCpuUsage',
+            expr: |||
+              sum by(node, job, cluster) (opensearch_os_cpu_percent) > %(alertsWarningCPUUsage)s
```
Comment (StefanKurek) [inline opensearch-mixin/alerts/alerts.libsonnet:108]:
```suggestion
              sum without(nodeid) (opensearch_os_cpu_percent) > %(alertsWarningCPUUsage)s
```

File: opensearch-mixin/alerts/alerts.libsonnet [hunk 13]
```diff
@@ -0,0 +1,205 @@
...
+              summary: 'The node CPU usage has exceeded the warning threshold.',
+              description: |||
+                {{$labels.opensearch_node}} has had {{ printf "%%.0f" $value }}%% CPU usage over the last 5m which is above the threshold of %(alertsWarningCPUUsage)s.
+              ||| % $._config,
+            },
+          },
+          {
+            alert: 'OpenSearchHighNodeCpuUsage',
+            expr: |||
+              sum by(node, job, cluster) (opensearch_os_cpu_percent) > %(alertsCriticalCPUUsage)s
```
Comment (StefanKurek) [inline opensearch-mixin/alerts/alerts.libsonnet:124]:
```suggestion
              sum without(nodeid) (opensearch_os_cpu_percent) > %(alertsCriticalCPUUsage)s
```

File: opensearch-mixin/alerts/alerts.libsonnet [hunk 14]
```diff
@@ -0,0 +1,205 @@
...
+              summary: 'The node CPU usage has exceeded the critical threshold.',
+              description: |||
+                {{$labels.opensearch_node}} has had {{ printf "%%.0f" $value }}%% CPU usage over the last 5m which is above the threshold of %(alertsCriticalCPUUsage)s.
+              ||| % $._config,
+            },
+          },
+          {
+            alert: 'OpenSearchModerateNodeMemoryUsage',
+            expr: |||
+              sum by(node, job, cluster) (opensearch_os_mem_used_percent) > %(alertsWarningMemoryUsage)s
```
Comment (StefanKurek) [inline opensearch-mixin/alerts/alerts.libsonnet:140]:
```suggestion
              sum without(nodeid) (opensearch_os_mem_used_percent) > %(alertsWarningMemoryUsage)s
```

File: opensearch-mixin/alerts/alerts.libsonnet [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+              summary: 'The node memory usage has exceeded the warning threshold.',
+              description: |||
+                {{$labels.opensearch_node}} has had {{ printf "%%.0f" $value }}%% memory usage over the last 5m which is above the threshold of %(alertsWarningMemoryUsage)s.
+              ||| % $._config,
+            },
+          },
+          {
+            alert: 'OpenSearchHighNodeMemoryUsage',
+            expr: |||
+              sum by(node, job, cluster) (opensearch_os_mem_used_percent) > %(alertsCriticalMemoryUsage)s
```
Comment (StefanKurek) [inline opensearch-mixin/alerts/alerts.libsonnet:156]:
```suggestion
              sum without(nodeid) (opensearch_os_mem_used_percent) > %(alertsCriticalMemoryUsage)s
```

File: opensearch-mixin/alerts/alerts.libsonnet [hunk 16]
```diff
@@ -0,0 +1,205 @@
...
+              summary: 'The node memory usage has exceeded the critical threshold.',
+              description: |||
+                {{$labels.opensearch_node}} has had {{ printf "%%.0f" $value }}%% memory usage over the last 5m which is above the threshold of %(alertsCriticalMemoryUsage)s.
+              ||| % $._config,
+            },
+          },
+          {
+            alert: 'OpenSearchModerateRequestLatency',
+            expr: |||
+              sum by(index, job, cluster) ((increase(opensearch_index_search_fetch_time_seconds{context="total"}[5m])+increase(opensearch_index_search_query_time_seconds{context="total"}[5m])+increase(opensearch_index_search_scroll_time_seconds{context="total"}[5m])) / clamp_min(increase(opensearch_index_search_fetch_count{context="total"}[5m])+increase(opensearch_index_search_query_count{context="total"}[5m])+increase(opensearch_index_search_scroll_count{context="total"}[5m]), 1)) > %(alertsWarningRequestLatency)s
```
Comment (StefanKurek) [inline opensearch-mixin/alerts/alerts.libsonnet:172]:
```suggestion
              sum without(context) ((increase(opensearch_index_search_fetch_time_seconds{context="total"}[5m])+increase(opensearch_index_search_query_time_seconds{context="total"}[5m])+increase(opensearch_index_search_scroll_time_seconds{context="total"}[5m])) / clamp_min(increase(opensearch_index_search_fetch_count{context="total"}[5m])+increase(opensearch_index_search_query_count{context="total"}[5m])+increase(opensearch_index_search_scroll_count{context="total"}[5m]), 1)) > %(alertsWarningRequestLatency)s
```

File: opensearch-mixin/alerts/alerts.libsonnet [hunk 17]
```diff
@@ -0,0 +1,205 @@
...
+              summary: 'The request latency has exceeded the warning threshold.',
+              description: |||
+                {{$labels.opensearch_index}} has had {{ printf "%%.0f" $value }}s of request latency over the last 5m which is above the threshold of %(alertsWarningRequestLatency)s.
+              ||| % $._config,
+            },
+          },
+          {
+            alert: 'OpenSearchModerateIndexLatency',
+            expr: |||
+              sum by(job, cluster, index) (increase(opensearch_index_indexing_index_time_seconds[5m]) / clamp_min(increase(opensearch_index_indexing_index_count[5m]), 1)) > %(alertsWarningIndexLatency)s
```
Comment (StefanKurek) [inline opensearch-mixin/alerts/alerts.libsonnet:188]:
```suggestion
              sum without(context) (increase(opensearch_index_indexing_index_time_seconds{context="total"}[5m]) / clamp_min(increase(opensearch_index_indexing_index_count{context="total"}[5m]), 1)) > %(alertsWarningIndexLatency)s
```

**Review by StefanKurek — CHANGES_REQUESTED** [review]

File: opensearch-mixin/dashboards/opensearch-cluster-overview.libsonnet [hunk 18]
```diff
@@ -0,0 +1,1614 @@
...
+      sort: 'none',
+    },
+  },
+};
+
+local topIndicesByIndexLatencyPanel = {
+  datasource: promDatasource,
+  targets: [
+    prometheus.target(
+      'topk(10, sum by(job, cluster, index) (increase(opensearch_index_indexing_index_time_seconds{job=~"$job", cluster=~"$opensearch_cluster"}[$__interval:]) / clamp_min(increase(opensearch_index_indexing_index_count{job=~"$job", cluster=~"$opensearch_cluster"}[$__interval:]), 1)))',
```
Comment (StefanKurek) [inline opensearch-mixin/dashboards/opensearch-cluster-overview.libsonnet:1386]:
```suggestion
      'topk(10, sum by(job, cluster, index) (increase(opensearch_index_indexing_index_time_seconds{job=~"$job", cluster=~"$opensearch_cluster", context="total"}[$__interval:]) / clamp_min(increase(opensearch_index_indexing_index_count{job=~"$job", cluster=~"$opensearch_cluster", context="total"}[$__interval:]), 1)))',
```

**Review by StefanKurek — CHANGES_REQUESTED** [review]

**Review by StefanKurek — APPROVED** [review]

File: opensearch-mixin/dashboards/opensearch-node-overview.libsonnet
Comment (Dasomeone) [inline opensearch-mixin/dashboards/opensearch-node-overview.libsonnet:1]: Some broken row elements here in this dashboard:
image
image

File: opensearch-mixin/dashboards/opensearch-search-and-index-overview.libsonnet
Comment (Dasomeone) [inline opensearch-mixin/dashboards/opensearch-search-and-index-overview.libsonnet:1]: Another broken row element here
image

**Review by Dasomeone — COMMENTED** [review]
Couple minor things and what appears as missing legends/labels on the overview dashboard, but this is otherwise looking fantastic! 

Last little thing, I didn't run the linting but please make sure `mixtool lint mixin.libsonnet` returns no issues (you'll need to vender in grafonnet first)

File: opensearch-mixin/dashboards/opensearch-cluster-overview.libsonnet [hunk 19]
```diff
@@ -0,0 +1,1614 @@
...
+  datasource: promDatasource,
+  targets: [
+    prometheus.target(
+      'topk(10, sort_desc(sum by(node, cluster, job) (opensearch_os_cpu_percent{job=~"$job", cluster=~"$opensearch_cluster"})))',
+      datasource=promDatasource,
+      legendFormat='{{node}}',
+    ),
+  ],
+  type: 'bargauge',
+  title: 'Top nodes by CPU usage',
```
[the remark this answers is not in the document]
Comment (JonathanWamsley) [reply opensearch-mixin/dashboards/opensearch-cluster-overview.libsonnet:327]: Hey Emily, I believe the panels are working correctly. For a single instance, it there are no legends shown. When I added the same query, then the legends appear just like in the sample app images shown above. Same applies for the Breakers tripped and Top nodes by disk usage panel, so I think this setup is good. Screenshot 2023-04-26 at 10 03 57 AM

I also updated the charts so now the row elements contain all the panels too.

**Comment by JonathanWamsley** [comment]
> Last little thing, I didn't run the linting but please make sure `mixtool lint mixin.libsonnet` returns no issues (you'll need to vender in grafonnet first)

No linting issues here
Screenshot 2023-04-26 at 10 12 33 AM

File: opensearch-mixin/dashboards/opensearch-cluster-overview.libsonnet (hunk 19 above)
[the remark this answers is not in the document]
Comment (Dasomeone) [reply opensearch-mixin/dashboards/opensearch-cluster-overview.libsonnet:327]: Ahh I see, that does explain it, thank you!

All looks good to me, thanks for all the hard work on this!

**Review by Dasomeone — APPROVED** [review]

File: opensearch-mixin/.lint [hunk 20]
```diff
@@ -0,0 +1,43 @@
+exclusions:
+  template-job-rule:
+    reason: "Prometheus datasource variable is being named as prometheus_datasource now while linter expects 'datasource'"
+  panel-datasource-rule:
+    reason: "Loki datasource variable is being named as loki_datasource now while linter expects 'datasource'"
+  template-datasource-rule:
+    reason: "Based on new convention we are using variable names prometheus_datasource and loki_datasource where as linter expects 'datasource'"
+  template-instance-rule:
+    reason: "Based on new convention we are using variable names prometheus_datasource and loki_datasource where as linter expects 'datasource'"
```
Comment (rgeyer) [inline opensearch-mixin/.lint:9]: FYI, soon, these won't be necessary any more 🎉 

grafana/dashboard-linter#124

**Review by rgeyer — APPROVED** [review]
LGTM, reviewers before me covered anything that stood out.

Thanks!

**Pull request merged**